### PR TITLE
Expect improvements to iosxr_setup.py US109031

### PR DESCRIPTION
**Pull request by RichWellum**
. Clean up send and wait expect logic. Comment is that I am currently throwing large chunks of configuration through the parser, while better expect behavior is to do one line by line checking the prompt.
. Remove funky multiple 'commits', 'sleeps' and 'end's.
. Remove all send runs, replace with one send run, commands then the send 'end'.
. ssh line gets rejected - needs logic to detect non k9 image and do the appropriate thing (show install active)
. Catch crypto or change crypto to add bits in so it isn’t working by accident.
. Remove 'pysun' naming throughout.

Tested with both k9 and non-k9 image. No errors from crypto now, and with crypto could still ssh in.

Box is generated successfully and basic testing looks good. Don't see the config lock issue anymore.

Still some nasty interaction with ZTP - temp. disabled it until it is fixed.

### CODE REVIEW

File: iosxr_setup.py [hunk 1]
```diff
@@ -33,49 +36,75 @@ def open_sessions(self, username, password):
...
         xr1.send(" port 57777")
-        xr1.send("telnet vrf default ipv4 server max-servers 10")
-        time.sleep(2)  # Give the parser a little time
+        xr1.wait("config-grpc")
+
+        # Commit changes and end
         xr1.send("commit")
-        time.sleep(2)  # Give the commit a little time
-        xr1.send("commit")  # Second commit helps when parser is holding onto config buffer
+        # A sleep and another commt can help if config locks are seen.
```
Comment (RichWellum) [inline iosxr_setup.py]: Will fix the typo. Keeping this code here commented out until I am sure it's stable.

File: iosxr_setup.py [hunk 2]
```diff
@@ -1,18 +1,21 @@
 '''
 Author: Rich Wellum (richwellum@gmail.com)
 
-Login to IOS XR, apply some config and then exit.
+Login to a runn IOS XR VM, apply a combination of XR and XR linux config to
```
Comment (glennmatthews) [inline iosxr_setup.py]: Login to a _running_ IOS XR VM, _then_ apply...

File: iosxr_setup.py [hunk 3]
```diff
@@ -33,49 +36,75 @@ def open_sessions(self, username, password):
...
         #
         # Configure xr
         #
+
+        # ZTP causes some startup issues so disable it during box creation
+        xr1.send("ztp terminate noprompt")
+
+        # Determine if the image is a crypto/k9 image or not
+        # This will be used to determine whether to configure ssh or not
+        xr1.send("run rpm -qa | grep k9sec")
```
Comment (glennmatthews) [inline iosxr_setup.py]: Do we want to get in the habit of using 'bash' instead of 'run'?

File: iosxr_setup.py [hunk 4]
```diff
@@ -33,49 +36,75 @@ def open_sessions(self, username, password):
...
+        time.sleep(2)
+        output = xr1.wait("[\$#]")
+        k9 = re.search(r'iosxrv-k9sec', output)
```
Comment (glennmatthews) [inline iosxr_setup.py]: For future compatibility with XRv 9000, maybe just `r'-k9sec'`?

File: iosxr_setup.py [hunk 5]
```diff
@@ -86,47 +115,57 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
         if allow_root_login:
             xr1.send("run sed -i 's/PermitRootLogin no/PermitRootLogin yes/' /etc/ssh/sshd_config_operns")
 
+        #
+        # Send commands to XR Linux
+        #
+        xr1.send("run")
```
Comment (glennmatthews) [inline iosxr_setup.py]: bash?

File: iosxr_setup.py [hunk 6]
```diff
@@ -86,47 +115,57 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
+        xr1.send("echo '%sudo ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers")
 
         # Add public key, so users can ssh without a password
         # https://github.com/purpleidea/vagrant-builder/blob/master/v6/files/ssh.sh
-        xr1.send("run [ -d ~vagrant/.ssh ] || mkdir ~vagrant/.ssh")
-        xr1.send("run chmod 0700 ~vagrant/.ssh")
-        xr1.send("run echo 'ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA6NF8iallvQVp22WDkTkyrtvp9eWW6A8YVr+kz4TjGYe7gHzIw+niNltGEFHzD8+v1I2YJ6oXevct1YeS0o9HZyN1Q9qgCgzUFtdOKLv6IedplqoPkcmF0aYet2PkEDo3MlTBckFXPITAMzF8dJSIFo9D8HfdOV0IAdx4O7PtixWKn5y2hMNG0zQPyUecp4pzC6kivAIhyfHilFR61RGL+GPXQ2MWZWFYbAGjyiYJnAmCP3NOTd0jMZEnDkbUvxhMmBYSdETk1rRgm+R4LOzFUGaHqHDLKLX+FIPKcF96hrucXzcWyLbIbEgE98OHlnVYCzRdK8jlqm8tehUc9c9WhQ== vagrant insecure public key' > ~vagrant/.ssh/authorized_keys")
-        xr1.send("run chmod 0600 ~vagrant/.ssh/authorized_keys")
-        xr1.send("run chown -R vagrant:vagrant ~vagrant/.ssh/")
+        xr1.send("[ -d ~vagrant/.ssh ] || mkdir ~vagrant/.ssh")
```
Comment (glennmatthews) [inline iosxr_setup.py]: Perhaps `mkdir -p ~vagrant/.ssh` instead?

File: iosxr_setup.py [hunk 7]
```diff
@@ -86,47 +115,57 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
         # /misc/app_host/scratch using the scp with user vagrant
         # E.g: scp -P 2200 Vagrantfile vagrant@localhost:/misc/app_host/scratch
-        xr1.send("run groupadd app_host")
-        xr1.send("run usermod -a -G app_host vagrant")
-        xr1.send("run mkdir /misc/app_host/scratch")
-        xr1.send("run chgrp -R app_host /misc/app_host/scratch")
-        xr1.send("run chmod 777 /misc/app_host/scratch")
+        xr1.send("groupadd app_host")
+        xr1.send("usermod -a -G app_host vagrant")
+        xr1.send("mkdir /misc/app_host/scratch")
```
Comment (glennmatthews) [inline iosxr_setup.py]: `mkdir -p` please

File: iosxr_setup.py [hunk 8]
```diff
@@ -86,47 +115,57 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
+        xr1.send("chgrp -R app_host /misc/app_host/scratch")
+        xr1.send("chmod 777 /misc/app_host/scratch")
 
         # Add Cisco OpenDNS IPv4 nameservers as a default DNS resolver
         # almost all users who have internet connectivity will be able to reach those.
         # This will prevent users from needing to supply another Vagrantfile or editing /etc/resolv.conf manually
-        xr1.send("run echo '# Cisco OpenDNS IPv4 nameservers' >> /etc/resolv.conf")
-        xr1.send("run echo 'nameserver 208.67.222.222' >> /etc/resolv.conf")
-        xr1.send("run echo 'nameserver 208.67.220.220' >> /etc/resolv.conf")
+        xr1.send("echo '# Cisco OpenDNS IPv4 nameservers' >> /etc/resolv.conf")
```
Comment (glennmatthews) [inline iosxr_setup.py:147]: Eventually going to move this to the XR CLI instead, right?

File: iosxr_setup.py [hunk 9]
```diff
@@ -86,47 +115,57 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
+        xr1.send("exit")
+        xr1.wait("RP/0/RP0/CPU0:ios")
 
-        # Generate a crypto key on XR. Note will fail if not a k9 image.
-        xr1.send("crypto key generate rsa")
+        # Set up IOS XR ssh if a k9/crypto image
+        if k9:
+            xr1.send("crypto key generate rsa")
+            xr1.wait("How many bits in the modulus")
+            xr1.send("")  # Send enter to get default 2048
```
Comment (glennmatthews) [inline iosxr_setup.py:168]: wait for prompt?

File: iosxr_setup.py (hunk 2 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 3 above)
Comment (RichWellum) [inline iosxr_setup.py]: Yes good point.

File: iosxr_setup.py (hunk 4 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 5 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 6 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 7 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 8 above)
Comment (RichWellum) [inline iosxr_setup.py:147]: Yes have opened a User Story for this.

File: iosxr_setup.py (hunk 9 above)
Comment (RichWellum) [inline iosxr_setup.py:168]: Ack.

File: iosxr_setup.py [hunk 10]
```diff
@@ -1,18 +1,21 @@
 '''
 Author: Rich Wellum (richwellum@gmail.com)
 
-Login to IOS XR, apply some config and then exit.
+Login to a running IOS XR VM, then apply a combination of XR and XR linux
```
Comment (RichWellum) [inline iosxr_setup.py:4]: @glennmatthews Fixed your comments. Retested both k9 and non-k9 images and works well.

**Comment by glennmatthews** [comment]
👍

**Comment by RichWellum** [comment]
Thanks!

||Rich

**Pull request merged**